bigquery/google/cloud/bigquery/dbapi/_helpers.py

```diff
@@ -0,0 +1,138 @@
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import collections
+import datetime
+import numbers
+import six
+import time
+
+from google.cloud import bigquery
+from google.cloud.bigquery.dbapi import exceptions
+
+
+def wait_for_job(job):
+    """Waits for a job to complete by polling until the state is `DONE`.
+
+    Raises a DatabaseError if the job fails.
+
+    :type: :class:`~google.cloud.bigquery.job._AsyncJob`
+    :param job: Wait for this job to finish.
+    """
+    while True:
+        job.reload()
+        if job.state == 'DONE':
+            if job.error_result:
+                # TODO: raise a more specific exception, based on the error.
+                # See: https://cloud.google.com/bigquery/troubleshooting-errors
+                raise exceptions.DatabaseError(job.errors)
+            return
+        time.sleep(1)
+
+
+def scalar_to_query_parameter(name=None, value=None):
+    """Convert a scalar value into a query parameter.
+
+    Note: the bytes type cannot be distinguished from a string in Python 2.
+
+    Raises a :class:`~ google.cloud.bigquery.dbapi.exceptions.ProgrammingError`
+    if the type cannot be determined.
+
+    For more information about BigQuery data types, see:
+    https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types
+
+    :type: str
+    :param name: Optional name of the query parameter.
+
+    :type: any
+    :param value: A scalar value to convert into a query parameter.
+
+    :rtype: :class:`~google.cloud.bigquery.ScalarQueryParameter`
+    """
+    parameter_type = None
+
+    if isinstance(value, bool):
+        parameter_type = 'BOOL'
+    elif isinstance(value, numbers.Integral):
+        parameter_type = 'INT64'
+    elif isinstance(value, numbers.Real):
+        parameter_type = 'FLOAT64'
+    elif isinstance(value, six.string_types):
+        parameter_type = 'STRING'
+    elif isinstance(value, bytes):
+        parameter_type = 'BYTES'
+    elif isinstance(value, datetime.datetime):
+        parameter_type = 'TIMESTAMP' if value.tzinfo else 'DATETIME'
+    elif isinstance(value, datetime.date):
+        parameter_type = 'DATE'
+    elif isinstance(value, datetime.time):
+        parameter_type = 'TIME'
+    else:
+        raise exceptions.ProgrammingError(
+            'encountered parameter {} with value {} of unexpected type'.format(
+                name, value))
+    return bigquery.ScalarQueryParameter(name, parameter_type, value)
+
+
+def to_query_parameters_list(parameters):
+    """Converts a list of parameter values into query parameters.
+
+    :type: list
+    :param parameters: List of query parameter values.
+
+    :rtype:
+        list of :class:`~google.cloud.bigquery._helpers.AbstractQueryParameter`
+    """
+    query_parameters = []
+
+    for value in parameters:
+        query_parameters.append(scalar_to_query_parameter(value=value))
+
+    return query_parameters
```

Comment:
(I think I didn't do this to begin with because I wanted to do something to support struct/list parameters, but I'd rather wait on that is it's a pretty unique-to-BigQuery feature. I wouldn't expect many people to try to pass in a dict as a singular parameter. Filed https://github.com/GoogleCloudPlatform/google-cloud-python/issues/3524